File: Logger/src/main/java/al/logger/client/ui/screens/music/Components/TextButton/TextButton.java
package al.logger.client.ui.screens.music.Components.TextButton;

import al.logger.client.ui.bases.ComponentBase;
import al.logger.client.utils.fontRender.ChineseFontRenderer;
import lombok.Getter;
import lombok.Setter;

import java.awt.*;

public class TextButton extends ComponentBase {

    @Getter
    @Setter
    private Runnable onClick;

    @Getter
    @Setter
    private String content;

    @Getter
    @Setter
    private ChineseFontRenderer renderer;

    @Getter
    @Setter
    private Color color;

    public TextButton(ChineseFontRenderer renderer, String content, float x, float y, Color color, Runnable onClick) {
        super("TextButton", x, y);
        this.onClick = onClick;
        this.renderer = renderer;
        this.content = content;
        this.color = color;
        float width = renderer.getStringWidth(content);
        float height = renderer.getHeight();
        this.getPosition().setWidth(width);
        this.getPosition().setHeight(height);
    }

    @Override
    protected void _initElements() {
    }

    @Override
    protected void _drawElement() {
        renderer.drawString(content, this.getPosition().getX(), this.getPosition().getY(), this.color.getRGB());
    }

    @Override
    protected boolean _mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if (isMouseHover) {
            onClick.run();
            return true;
        }
        return false;
    }

    @Override
    protected boolean _mouseReleased(int mouseX, int mouseY, int mouseStatus) {
        return false;
    }

    @Override
    public void _mouseDWheel(int mouseX, int mouseY, int mouseDWheel) {
    }

    @Override
    public void update() {
    }

    @Override
    public void keyTyped(char p_keyTyped_1_, int p_keyTyped_2_) {
    }
}

File: Logger/LoggerInjector/examples/example_win32_opengl3/TextButton.cpp
#include "TextButton.h"
#include "Fonts.h"
#include "render.h"

bool TextButton::MouseDWheel(float mouseX, float mouseY, int mouseDWheel)
{
	return false;
}

void TextButton::update()
{
	ImFont* geologicaRegular18 = FontsLoader::GetGeologicaRegular18();
	ImVec2 textSize = geologicaRegular18->CalcTextSizeA(geologicaRegular18->FontSize, FLT_MAX, 0, this->text.c_str());
	this->position.setWidth(textSize.x);
	this->position.setHeight(textSize.y);
}

void TextButton::update(ComponentBase* parent)
{
	update();
}

void TextButton::KeyTyped(int code, wchar_t in_char)
{
}

void TextButton::DrawElement()
{
	ImFont* geologicaRegular18 = FontsLoader::GetGeologicaRegular18();
	alphaAnimation.update(0.2, isMouseHover);
	RenderUtil::drawText(this->position.getX(), this->position.getY(), this->text, ImColor(30, (int)(140 + alphaAnimation.getValue()), 255), geologicaRegular18);
}

bool TextButton::MouseClicked(float mouseX, float mouseY, int mouseButton)
{
	if (isMouseHover) {
		this->MouseClickedLambda();
		return true;
	}
	return false;
}

bool TextButton::MouseReleased(float mouseX, float mouseY, int mouseStatus)
{
	return false;
}

TextButton::TextButton(std::string text, std::function<void()> MouseClickedLambda):ComponentBase("TextButton"),text(text), MouseClickedLambda(MouseClickedLambda)
{
}
